src/core/openjpeg_decoder.js

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BaseException } from "../shared/util.js";
+import OpenJPEG from "../../external/openjpeg/openjpeg.js";
+
+class JpxError extends BaseException {
+  constructor(msg) {
+    super(`JPX error: ${msg}`, "JpxError");
+  }
+}
+
+class JpxImage {
+  #module = null;
+
+  static #instance = null;
+
+  constructor(module) {
+    this.#module = module;
+  }
+
+  static decode(data) {
+    this.#instance ||= new JpxImage(OpenJPEG());
+    return this.#instance.#decode(data);
+  }
+
+  static cleanup() {
+    this.#instance = null;
+  }
+
+  #decode(data) {
+    const module = this.#module;
+    const data_ptr = module._malloc(data.length);
```

Comment:
I'm not sure how feasible this is, but figured I'd ask just in case: would it be possible to provide a wrapper function in e.g. https://github.com/mozilla/pdf.js.openjpeg/blob/main/src/decoder.c that does most of this function's logic, but then in C? Looking at the code in this method it feels a bit like C details like explicit malloc/free are leaking through to JavaScript, and I'm not sure if that's wanted or not. If we could expose a `jpx_decode(data)` API from `decoder.c` we might be able to move all those details into the C code, which would simplify the JavaScript code here and keep e.g. malloc/free concerns in C where they are relevant.